extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+full power of XUTXO-3 and so could use a simpler model, perhaps with
+less computational overhead.
+
+\bigskip
+
+\noindent Before presenting the extended models we will define some
+basic notation and give a little background on the standard UTXO
+model.
+
+
+\section{Notation}
+This section defines some basic notation.  We generally follow the
+notation established by \cite{Zahnentferner18-UTxO}.
+
+\begin{ruledfigure}{H}
+  \begin{itemize}
+\item Types are typeset in $\mathsf{sans~serif}$.
+
+\item All scripts are of type $\mathsf{Script}$.
+
+\item The only operation on $\mathsf{Script}$ from the ledger's
+  perspective is applying a validator script to three arguments,
+  which is denoted by $\llbracket \cdots \rrbracket :
+  \mathsf{Script} \rightarrow \mathsf{Script} \times \mathsf{Script}
+    \times \mathsf{Script} \rightarrow \mathbb{V}$ where $\mathbb{V}$
+    is some type containing a distinguished value
+    \textsf{error}\footnotemark.
+
+\item A record type with fields $\phi_1, \ldots, \phi_n$ of types $T_1,
+  \ldots, T_n$ is denoted by $(\phi_1 : T_1, \ldots, \phi_n : T_n)$.
+
+\item If $t$ is a value of a record type $T$ and $\phi$ is the name
+  of a field of $T$ then $t.\phi$ denotes the value of $\phi$ for
+  $t$.
+  
+\item A list $\lambda$ of type $\List{T}$ is
+    either the empty list $[]$ or a list $e :: \lambda'$ with $head$
+    $e$ of type $T$ and $tail$ $\lambda'$ of type
+    $\List{T}$. We denote the $i$th element of a list
+    $L$ by $L[i]$.
+    The concatenation of two lists $\lambda_1$ and
+    $\lambda_2$ is denoted $\lambda_1 ::: \lambda_2$.
+
+  \item For two types $K$ and $V$, $\Map{K}{V}$ denotes the type of
+    partial functions from \textit{keys} in $K$ to \textit{values} in
+    $V$.  If $m$ is such a map then $\dom m$ (the \textit{domain} of
+    $m$) is the set of keys of $m$ (ie, $K$).  Given a map $m$ and a
+    key $k \in \dom m$, we denote the result of looking up $k$ in $m$
+    by $m[k]$.  Equality for maps is defined in the obvious way: $m_1
+    = m_2$ if and only if $\dom m_1 = \dom m_2$ and $m_1[k] = m_2[k]$
+    for all $k \in \dom m_1$ (assuming we have decidable equality in $V$).
+    
+  \item If the type $V$ is a monoid with binary operation $+$ (for
+    example, a numeric type) and $f, g \in \Map{K}{V}$ then we define
+    their sum $f+g \in \Map{K}{V}$ with domain $\dom (f+g) = \dom f
+    \cup \dom g$ by
+    \[
+    (f+g)[k] =
+    \left\{ \begin{array}{ll}
+        f[k] + g[k] & \mbox{if $k \in \dom f \cap \dom g$}\\
+        f[k] & \mbox{if $k \in \dom f \setminus \dom g$}\\
+        g[k] & \mbox{if $k \in \dom g \setminus \dom f$}
+      \end{array}
+      \right.
+      \]
+      We generalise this to the sum of a finite number of maps using
+      the $\sum$ notation in the usual way.  Note that the type
+      $\Map{K}{V}$ itself becomes a monoid under the operation $+$
+      which we have just defined (with the empty map as identity element).
```

Comment:
Sort of: adding `{}` will leave a map completely unchanged,  whereas adding `{a:0}` may extend the domain, so they're slightly different (the current definition of domain is wrong: what I intended it to be is the set of keys contained in the map).  You're right though: we do need to be more careful about equality because your two examples are the same in some sense.  Can we get this in practice?  I'll have to think about that.

Reply:
Yes, `{a : 1} - {a :1} = {a :0}`. `x-x` should be the identity, so that had better be the same as the empty map.

Reply:
The problem is that there's an existing well-defined notion of equality for maps, and this isn't it.  Instead, we've got an equivalence relation on maps and we really need to work with the quotient (and formalisers always tell me those are unpleasant), and show, eg, that the relation's a congruence wrt +.  There's also the problem that we have to propagate this stuff through another layer of maps for the native NFT scheme, which will get messy.

I think the easiest way to deal with this is to stick with the pointwise definition of +, which lifts the  property of being a monoid (but not a group!) from V to Map[K,V], and define a pruning operation which throws away the things that map to 0.  This gives us a canonical representative of each equivalence class, and we only have to worry about equality for those.  This means that we can state validity properties by saying that if you prune both sides you get things that really are equal.